### ISSUE TITLE
[3.0] massive fragmentation on NTFS

### ISSUE THREAD

**Issue by Tharn** [transmission/transmission#869]
Using 64-bit transmission-daemon on Windows 8.1, the default setting of `preallocation=1` results in massive fragmentation even on empty volumes with loads of free contiguous space. NTFS supports sparse files, so what gives? Is this the intended result?

What I would expect: `Preallocation=1` writes the first and the last chunk of a file. It should properly reserve the space in between and fill in the chunks as they come. The end result should be coherent files.

What actually happens:
![Untitled](https://user-images.githubusercontent.com/13466852/55350948-7a5d5580-54bd-11e9-9e92-ab0900df965d.png)

![Untitled2](https://user-images.githubusercontent.com/13466852/55351508-d379b900-54be-11e9-9f1d-3de02c1d9f00.png)

As you can see, the partition is virtually empty except for the torrent. All the yellow color is the result of fragmented files from a single torrent.

The long-term result:
When even an almost empty partition is filled by highly fragmented files such as these, there quickly comes a point when the NTFS filesystem exhausts its pool of 'free fragments'. The limit is high, but yes these are limited. You saw how many fragments 3 single files managed to accrue. When this happens, NTFS starts to create weird overflow files (the hidden ones with $ in the name). I had this happen some months ago on another harddisk. You cannot get rid of these to repair your filesystem unless you forcibly move content off the disk and back - or delete it.

Why not `preallocation=2` ?
This is fine as long as you're using a HDD. This setting writes data to disk twice. First it fills the entirety of the space used by the torrent with zeroes, then it overwrites those zeroes with data. So it spends twice the amount of writes. Using an SSD, you want to minimize writes. Especially with large media files.

What happens with `preallocation=0`?
Strangely enough, this setting produces very little fragmentation. Large media files are usually shown in 2-3 fragments as opposed to hundreds of them. Makes me wonder if we should set this as the Windows default instead.

**Comment by msg7086** [comment]
A quick thought. Did you run transmission under Admin mode?

**Comment by Tharn** [comment]
Prealloc hasn't changed for new version, problem still exists with `prealloc=1` trashing NTFS after a while.

**Comment by LackThereof** [comment]
I don't think this is necessarily a fixable Transmission bug — this is just Windows behavior when you ask it to create a sparse file on NTFS, and then write to random parts of it. Windows will not reserve physical space on the disk for any significantly-sized portion of the file that hasn't be written to yet. There's nothing I'm aware of that Transmission can do about it, other than actually write out the whole file sequentially (`preallocation=2`).  I would of course love to be shown that my belief is wrong.

You may ask
> Why would windows do this?

Well without banging on doors in Redmond we can't have the full picture, but there are a couple educated guesses we can make.
* To save disk space
In this way, we can successfully create many files exceeding the capacity of the actual physical disk, and it becomes easier to generally avoid disk space crunches. This may seem silly to those of us who will be rapidly filling those large sparse files over a period of hours or days. But to, say, a virtualization app creating virtual disks which may not become full any time soon (if ever), it can be handy.
This space advantage is actually touted in Windows documentation as the headline benefit of sparse files.
* The MS folks determined way back when that the average sequential read performed was actually a relatively small size (in comparison to, say, large media files) _n_, and that so long as the fragments of a fragmented file were significantly larger than that small size _n_, real-world performance impacts of fragmentation were minimal and not worth the effort of defragmentation. If my memory serves, they used some similar explanation when asked about changes to the built-in defragmentation tool around the turn of the century or so.

**Comment by msg7086** [comment]
I completely forgot that I replied to this thread.

NTFS supports preallocation through `SetFileValidData`. It does require privilege to use (because it may potentially expose existing data on disk). Preallocation through sparse file is not really a preallocation (sparse file doesn't allocate any space).

**Comment by Tharn** [comment]
It's fair enough that this is something Windows does. That then makes sparse files ununsable for a torrent client on windows because thousands of fragments *per iso-sized torrent* is neither acceptable nor usable. It trashes the filesystem in quick order. The default value needs to change if prealloc itself cannot be changed .
